### ISSUE TITLE
Remote Debugging OS/161

### ISSUE THREAD

**Issue by rishabh** [microsoft/vscode-cpptools#682]
Hi everyone,

First off, thank you all for your hard work!

## System:
- OS: macOS Sierra 10.12
- VS Code: 1.11.2
- vscode-cpptools: 0.11.0
- gdb: GNU gdb 6.6

## The Problem
1. Breakpoints do not work
   * Attempting to set breakpoints in VS Code returns "Attempting to bind the breakpoints...."
   * `"stopAtEntry": true` does not stop execution
2. Pause button does not do anything while "debugging"

## Some Background Info
I'm currently trying to debug an OS/161 kernel using gdb, through VS Code.  Unlike most programs, OS/161 is set up like so:

```bash
# Set up the sys161 simulator using the kernel,
#  and wait until gdb has connected to unix:.sockets/gdb
$> sys161 -w kernel

# Connect to unix:.sockets/gdb
$> gdb kernel
(gdb) target remote unix:.sockets/gdb
```
I've managed to connect to the UNIX socket following the examples posted here, but I cannot get gdb to properly integrate. My `launch.json` looks like this:
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug",
            "type": "cppdbg",
            "request": "launch",
            "preLaunchTask": "Compile",
            "program": "~/os161/root/kernel",
            "cwd": "~/os161/root",
            "miDebuggerServerAddress": "unix:.sockets/gdb",
            "miDebuggerPath": "/usr/local/bin/gdb",
            "launchCompleteCommand": "exec-continue",
            "stopAtEntry": true,
            "logging": { "trace": true, "traceResponse": true }
        }
    ]
}
```
## Logs

At start:
```json
R: {"success":true,"message":null,"request_seq":5,"command":"threads","body":{"threads":[]},"running":false,"refs":null,"seq":0,"type":"response"}
```

Placing a breakpoint:
```json
C setBreakpoints: {"source":{"path":"/Users/user/os161/os161-1.99/root/kern/startup/menu.c"},"lines":[77],"breakpoints":[{"line":77}],"sourceModified":false}
R: {"success":true,"message":null,"request_seq":6,"command":"setBreakpoints","body":{"breakpoints":[{"id":1,"verified":true,"line":77,"message":null}]},"running":false,"refs":null,"seq":0,"type":"response"}
E breakpoint: {"reason":"changed","breakpoint":{"id":1,"verified":false,"line":77,"message":"Attempting to bind the breakpoint...."},"type":"breakpoint"}
```
Any help would be appreciated, thank you!

### Edit:
~Appears that gdb 6.6 does not support GDB/MI so it doesn't look like this'll work with VS Code.~
Appears that gdb 6.6 DOES support GDB/MI, the issue is that gdb 6.6 does not support ``-thread-info`` and/or ``-gdb-set target async [on/off]``.

**Comment by pieandcakes** [comment]
@FlapperDoodlez  We use the MI interpreter to talk to the debugger so unless it supports MI I don't think it will work. You can try and do remote through gdbserver if your local version of gdb speaks MI that *might* be a workaround. To see the MI commands we send, you can enable `"logging": { "engineLogging": true }`

**Comment by rishabh** [comment]
@pieandcakes Thank you for your reply!
I set up gdbserver on a remote machine and managed to connect to it. Unfortunately, running `target remote unix:.sockets/gdb` causes the local gdb to disconnect from the remote gdbserver.

It seems impossible at this point.

Since I'm not really happy with the current frontends for gdb version < 7.0, I think I'm just going to make my own frontend for gdb version < 7.0.

It doesn't look like this issue is related to VS Code anymore so I'm going to close it, but thanks again!

**Comment by rishabh** [comment]
@pieandcakes I think I may have jumped the gun on this one.

Looks like GDB/MI **is** supported. The issue is that nothing really happens in VSCode after I connect. Here's what the Debug Console outputs after I try to continue execution:
```
1: (278) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'
1: (291) LaunchOptions  ExePath='~/os161/root/kernel'
1: (291) LaunchOptions  WorkingDirectory='~/os161/root'
1: (291) LaunchOptions  ExeArguments=''
1: (291) LaunchOptions  MIDebuggerPath='/usr/local/bin/gdb'
1: (291) LaunchOptions  WaitDynamicLibLoad='false'
1: (291) LaunchOptions  MIDebuggerServerAddress='unix:.sockets/gdb'
1: (291) LaunchOptions>
1: (291) LaunchOptions    <LaunchCompleteCommand>exec-continue</LaunchCompleteCommand>
1: (292) LaunchOptions</LocalLaunchOptions>
1: (468) Starting: "/usr/local/bin/gdb" --interpreter=mi
1: (498) DebuggerPid=22913
1: (513) ->~"GNU gdb 6.6\n"
1: (519) ->~"Copyright (C) 2006 Free Software Foundation, Inc.\n"
1: (519) ->~"GDB is free software, covered by the GNU General Public License, and you are\n"
1: (519) ->~"welcome to change it and/or distribute copies of it under certain conditions.\n"
1: (519) ->~"Type \"show copying\" to see the conditions.\n"
1: (519) ->~"There is absolutely no warranty for GDB.  Type \"show warranty\" for details.\n"
1: (519) ->~"This GDB was configured as \"--host=i386-apple-darwin16.0.0 --target=mips-harvard-os161\".\n"
1: (519) ->&"No symbol table is loaded.  Use the \"file\" command.\n"
1: (519) ->(gdb)
1: (530) <-1001-gdb-set target-async on
1: (531) ->&"No symbol table is loaded.  Use the \"file\" command.\n"
1: (531) ->1001^error,msg="No symbol table is loaded.  Use the \"file\" command."
1: (531) ->(gdb)
1: (534) 1001: elapsed time 5
1: (545) <-1002-gdb-set solib-search-path ~/os161/root:
1: (546) ->1002^done
1: (546) ->(gdb)
1: (546) 1002: elapsed time 0
1: (546) <-1003-environment-cd ~/os161/root
1: (546) ->1003^done
1: (546) ->(gdb)
1: (546) 1003: elapsed time 0
1: (546) <-1004-file-exec-and-symbols ~/os161/root/kernel
1: (551) ->1004^done
1: (551) ->(gdb)
1: (551) 1004: elapsed time 5
1: (555) <-1005-interpreter-exec console "show architecture"
1: (555) ->~"The target architecture is set automatically (currently mips:3000)\n"
1: (556) ->1005^done
1: (556) ->(gdb)
1: (556) 1005: elapsed time 1
1: (563) <-1006-break-insert main
1: (563) ->&"Function \"main\" not defined.\n"
1: (564) ->1006^done
1: (564) ->(gdb)
1: (564) 1006: elapsed time 0
1: (564) <-1007-target-select remote unix:.sockets/gdb
1: (566) ->~"Current language:  auto; currently asm\n"
1: (566) ->1007^connected,thread-id="0",frame={addr="0xffffffff80029f50",func="__start",args=[],file="../../arch/sys161/startup/start.S",line="54"}
1: (566) ->(gdb)
1: (567) 1007: elapsed time 3
GNU gdb 6.6
Copyright (C) 2006 Free Software Foundation, Inc.
GDB is free software, covered by the GNU General Public License, and you are
welcome to change it and/or distribute copies of it under certain conditions.
Type "show copying" to see the conditions.
There is absolutely no warranty for GDB.  Type "show warranty" for details.
This GDB was configured as "--host=i386-apple-darwin16.0.0 --target=mips-harvard-os161".
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
Current language:  auto; currently asm
c
Unable to perform this action because the process is running.
```

I'm not sure where to go from here so I'm going to reopen the issue because I think it may actually be VS Code related.

Edit: It also looks like this version of gdb doesn't support ``-thread-info`` and since VS Code relies on ``-thread-info``, this doesn't look like it's going to work -- for real this time.

**Comment by pieandcakes** [comment]
@FlapperDoodlez I think another issue will be : 
```
1: (556) 1005: elapsed time 1
1: (563) <-1006-break-insert main
1: (563) ->&"Function \"main\" not defined.\n"
``` 
We rely on the initial breakpoint to setup the debugging environment. You can play with `customLaunchSetupCommands` if you can get `gdb` to work by passing a list of setup commands that aren't the ones we use but I suspect without `-thread-info` it probably won't behave well either. I've asked other users in the past that have gotten a similar scenario working to try and contribute documentation but I haven't seen anything come through. if you search through open/closed issues, you might be able to connect with one of them to see what they have done to work around it.

**Comment by rishabh** [comment]
@pieandcakes Good news and bad news!

### Good news:
I found out that the gdb that works with OS/161 was updated by Harvard to v7.8. I installed it from here http://os161.eecs.harvard.edu/download/
I've actually been able to connect (properly this time) to gdb from VS Code.

### Bad news:
You're right about the breakpoint at main. There actually isn't a ``main`` function to connect to in the kernel so an error is displayed, but it's not a _big_ deal since the error is just skipped over. But it may be related to a new bug I've come across:

For some reason, the debugger is sending an ``-exec-continue`` even when I've explicitly told it not to.
Here is my ``launch.json``:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug",
            "type": "cppdbg",
            "request": "launch",
            //"preLaunchTask": "Compile",
            "program": "~/os161/root/kernel",
            "cwd": "~/os161/root",
            "miDebuggerServerAddress": "unix:.sockets/gdb",
            "miDebuggerPath": "/usr/local/bin/gdb",
            "launchCompleteCommand": "None",
            "MIMode": "gdb",
            "logging": {
                "engineLogging": true
            }
        }
    ]
}
```

And here is my console output, the ``-exec-continue`` command is at the very bottom:
```
...
1: (556) <-1006-break-insert main
1: (557) ->1006^error,msg="Function \"main\" not defined."
1: (557) ->(gdb)
1: (558) 1006: elapsed time 2
1: (558) <-1007-target-select remote unix:.sockets/gdb
1: (559) ->=thread-group-started,id="i1",pid="42000"
1: (564) ->=thread-created,id="1",group-id="i1"
1: (565) ->~"__start () at ../../arch/sys161/startup/start.S:54\n"
1: (565) ->&"54\t../../arch/sys161/startup/start.S: No such file or directory.\n"
1: (565) ->*stopped,frame={addr="0x80029f50",func="__start",args=[],file="../../arch/sys161/startup/start.S",fullname="../../arch/sys161/startup/start.S",line="54"},thread-id="1",stopped-threads="all"
1: (565) ->1007^connected
1: (565) ->(gdb)
1: (575) <-1008-thread-info 1
1: (583) ->1008^done,threads=[{id="1",target-id="Thread <main>",frame={level="0",addr="0x80029f50",func="__start",args=[],file="../../arch/sys161/startup/start.S",fullname="../../arch/sys161/startup/start.S",line="54"},state="stopped"}]
1: (583) ->(gdb)
1: (593) 1007: elapsed time 34
1: (598) 1008: elapsed time 22
=thread-group-added,id="i1"
GNU gdb (GDB) 7.8
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "--host=x86_64-apple-darwin16.0.0 --target=mips-harvard-os161".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
__start () at ../../arch/sys161/startup/start.S:54
1: (667) <-1009-interpreter-exec console "info sharedlibrary"
1: (667) ->~"No shared libraries loaded at this time.\n"
1: (668) ->1009^done
1: (668) ->(gdb)
1: (669) 1009: elapsed time 2
Loaded 'shared libraries loaded at this time.'. Cannot find or open the symbol file.
1: (687) <-1010-thread-info
1: (687) ->1010^done,threads=[{id="1",target-id="Thread <main>",frame={level="0",addr="0x80029f50",func="__start",args=[],file="../../arch/sys161/startup/start.S",fullname="../../arch/sys161/startup/start.S",line="54"},state="stopped"}],current-thread-id="1"
1: (687) ->(gdb)
1: (689) 1010: elapsed time 2
1: (703) <-1011-stack-list-frames 0 1000
1: (704) ->1011^done,stack=[frame={level="0",addr="0x80029f50",func="__start",file="../../arch/sys161/startup/start.S",fullname="../../arch/sys161/startup/start.S",line="54"}]
1: (704) ->(gdb)
1: (706) 1011: elapsed time 3
1: (712) <--exec-continue
1: (712) ->^running
1: (713) ->*running,thread-id="all"
1: (713) ->(gdb)
1: (713) ->(gdb)
```
I've managed to work around this by manually setting a breakpoint on the entrypoint in the "Breakpoints" pane **before** starting the debugger, that way I can still interact with gdb.

* * *

Another bug I've experienced is that if I set the ``customLaunchSetupCommands`` to anything (even empty), I'll get an error message at the top of VS Code saying "Cannot read property 'toLowerCase' of undefined" and get this at the bottom of my output:
```
...
1: (489) ->~"Type \"apropos word\" to search for commands related to \"word\".\n"
1: (489) ->(gdb)
1: (501) <-1001-gdb-set target-async on
1: (502) ->1001^done
1: (502) ->(gdb)
1: (504) 1001: elapsed time 3
1: (526) <-logout
1: (527) ->&"logout\n"
```
Trying to run the same configuration causes the same error message to pop up, the debugger doesn't even start. I've worked around this bug only by restarting VS Code (and also not using ``customLaunchSetupCommands``).

**Comment by pieandcakes** [comment]
I think you want `"customLaunchSetupCommands"` as that overrides what we do. Also, since it expects it to break at `main()` because of a breakpoint we set, that is what the `-exec-continue` is for.

**Comment by rishabh** [comment]
Thank you for the quick reply! I think you're right, ``customLaunchSetupCommands`` would probably solve the issues I'm having.

The issue is that I get a weird error when I use ``customLaunchSetupCommands`` saying "Cannot read property 'toLowerCase' of undefined" (none of the launch commands run).

My launch.json:
```
...
"launchCompleteCommand": "None",
"customLaunchSetupCommands": [
        {
            "text": "-environment-cd ~/os161/root"
        },
        {
            "text": "-file-exec-and-symbols ~/os161/root/kernel"
        },
        {
            "text": "-interpreter-exec console \"show architecture\""
        },
        {
            "text": "-target-select remote unix:.sockets/gdb"
        }
    ],
"MIMode": "gdb",
...
```

Errors I get in the console in Developer Tools:
![screen shot 2017-05-05 at 10 09 16 pm](https://cloud.githubusercontent.com/assets/7551150/25769100/855ebe18-31df-11e7-9eac-ec619a0bebb9.png)

**Comment by pieandcakes** [comment]
@FlapperDoodlez That's strange. Can you send me the entire launch.json block you are sending? I wonder what its calling toLowerCase on?

**Comment by rishabh** [comment]
@pieandcakes Sure!

It doesn't really matter what I put in `"customLaunchSetupCommands"`, it'll give me a "toLowerCase" error. I've even set customLaunchSetupCommands to an empty array, it'll still give me that error.

It's interesting to note that when I get that error even once, I will _instantly_ get that error again when I start to debug so nothing even runs.

Here is my launch.json:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug",
            "type": "cppdbg",
            "request": "launch",
            "program": "~/os161/root/kernel",
            "cwd": "~/os161/root",
            "stopAtEntry": false,
            "miDebuggerServerAddress": "unix:.sockets/gdb",
            "miDebuggerPath": "/usr/local/bin/gdb",
            "launchCompleteCommand": "None",
            "MIMode": "gdb",
            "customLaunchSetupCommands": [
                {
                    "text": "-file-exec-and-symbols ~/os161/root/kernel"
                }
            ],
            "logging": {
                "engineLogging": true
            }
        }
    ]
}
```

**Comment by pieandcakes** [comment]
@FlapperDoodlez I wonder if some other extension is throwing that error? I put that block in my VSCode 1.12.1 with latest 0.11.0 of the extension and I don't see the toLowerCase() error pop up. The toLowerCase() error happens before our debugger code gets it so it should be duplicatible.

**Comment by rishabh** [comment]
@pieandcakes Thank you for your quick reply!

I disabled all my extensions except cpptools, and I am running VS Code 1.12.1 with cpptools 0.11.0.

I think I may have left out an important piece of info. I'm not sure about the error in which I get the messages, but I may be getting "Debug adapter process has terminated unexpectedly" _first_ and _then_ getting "Cannot read property 'toLowerCase' of undefined". Subsequent attempts at debugging causes me to receive the "Cannot read property 'toLowerCase' of undefined" error and _not_ the "Debug adapter ..." error.

What's also interesting is that the custom commands don't even run, I get this output when I turn on engine logging:
```
...
1: (534) ->(gdb)
1: (540) <-1001-gdb-set target-async on
1: (541) ->1001^done
1: (541) ->(gdb) // my commands should run here
1: (543) 1001: elapsed time 4 
1: (565) <-logout
1: (565) ->&"logout\n"
```
---
I'm going to do a clean install of VS Code + only cpptools tomorrow and see if I get the error.

**Comment by rishabh** [comment]
@pieandcakes I did a clean install and I still get the error. If it's not being caused by cpptools, I may make an issue on VS Code github repo.

I've just worked around it by placing a breakpoint before starting gdb. Thanks again for your help!

**Comment by pieandcakes** [comment]
@FlapperDoodlez Is this happening with local debugging too?

**Comment by pieandcakes** [comment]
@FlapperDoodlez and apparently VSCode says its an extension issue.

**Comment by pieandcakes** [comment]
https://github.com/Microsoft/vscode/issues/25042
https://github.com/Microsoft/vscode/issues/26374

**Comment by pieandcakes** [comment]
@FlapperDoodlez Looks like if you update to the new VSCode Insiders, the toLowerCase issue should be fixed. 
https://github.com/Microsoft/vscode/issues/26374

**Comment by pieandcakes** [comment]
@FlapperDoodlez if you want to take a stab at debugging the scenario, I've included instructions to debug the MIEngine tier if you want to try it out for me.

https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/Debugger/How%20To%20Debug%20MIEngine.md

**Comment by WardenGnaw** [comment]
Closing issue with no recent updates. 

If this issue is still occuring, please open a new issue.
